(feat): improve-multilingual-support

narsimhaReddyJuspay opened the pull request:
- added dynamic tts language switching
- address based language selection for stt and tts
- improved prompt for better multilingual support

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Added comprehensive multi-language support for voice agents with automatic language detection from request payloads
  * Implemented language-aware text-to-speech service with intelligent script detection for automatic language switching
  * Introduced template configuration options to enable payload-based language selection for calls

* **Configuration**
  * Added new runtime configuration option to manage text-to-speech preprocessing behavior

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

swaroopvarma1 commented on app/ai/voice/agents/breeze_buddy/utils/language.py:
```diff
@@ -0,0 +1,202 @@
+    if not address:
+        return None
+
+    return await extract_state_with_llm(address)
```
we should not use LLM here

narsimhaReddyJuspay commented on app/ai/voice/agents/breeze_buddy/agent.py:
```diff
@@ -220,6 +226,23 @@ async def run(self):
 
+        # Add language_name to template_vars for prompt customization
+        # Language detection is controlled by payload_based_language_selection flag in template
+        detected_language = call_payload.get("detected_language")
```
detected_language_code

narsimhaReddyJuspay commented on app/ai/voice/agents/breeze_buddy/utils/language.py:
```diff
@@ -0,0 +1,201 @@
+Language code:"""
+
+    try:
+        async with httpx.AsyncClient(timeout=10.0) as client:
```
.

narsimhaReddyJuspay commented on app/ai/voice/agents/breeze_buddy/agent.py:
```diff
@@ -88,6 +90,10 @@ def __init__(
         self.expected_callback_response_schema = None
+        
+        # Language and TTS config (initialized during run)
+        self.is_sarvam = False
```
.

narsimhaReddyJuspay commented on app/ai/voice/agents/breeze_buddy/agent.py:
```diff
@@ -220,6 +226,23 @@ async def run(self):
+        self.template_vars["primary_language"] = self.language_name
+
+        # Check TTS service for prompt customization
+        tts_service = await BB_TTS_SERVICE()
```
.

narsimhaReddyJuspay commented on app/ai/voice/agents/breeze_buddy/agent.py:
```diff
@@ -220,6 +226,23 @@ async def run(self):
+
+        # Check TTS service for prompt customization
+        tts_service = await BB_TTS_SERVICE()
+        self.is_sarvam = tts_service.lower() == "sarvam"
```
.

swaroopvarma1 commented on app/ai/voice/agents/breeze_buddy/utils/language.py:
```diff
@@ -0,0 +1,202 @@
+    Returns:
+        List of language hints for STT (e.g., ["te", "en"])
+    """
+    default_hints = ["hi", "en"]
```
en, hi

swaroopvarma1 commented on app/ai/voice/agents/breeze_buddy/examples/templates/order-confirmation-2.json:
```diff
@@ -64,7 +65,7 @@
             "role": "system",
-            "content": "You are Rhea, a friendly human and natural-sounding customer care representative from {shop_name}. Your goal is to confirm a Cash on Delivery (COD) order with the customer.\n\nIMPORTANT: You are a text-to-speech voice agent. Generate all responses as plain, unformatted text without any special characters, markdown formatting, asterisks, underscores, or other unnecessary sy ...
+            "content": "You are Rhea, a friendly human and natural-sounding customer care representative from {shop_name}. Your goal is to confirm a Cash on Delivery (COD) order with the customer.\n\nIMPORTANT: You are a text-to-speech voice agent. Generate all responses as plain, unformatted text without any special characters, markdown formatting, asterisks, underscores, or other unnecessary sy ...
           }
```
.

swaroopvarma1 commented on app/ai/voice/agents/breeze_buddy/stt/__init__.py:
```diff
@@ -54,13 +56,20 @@ async def get_stt_service(language_hints: str | None = None):
         bb_sarvam_stt_vad_signals = await BB_SARVAM_STT_VAD_SIGNALS()
         bb_sarvam_stt_high_vad_sensitivity = await BB_SARVAM_STT_HIGH_VAD_SENSITIVITY()
 
+        # Use detected language if provided, otherwise fall back to config
```
.

swaroopvarma1 commented on app/ai/voice/agents/breeze_buddy/utils/language.py:
```diff
@@ -0,0 +1,202 @@
+}
+
+# Language code to STT hints (for Soniox)
+LANGUAGE_TO_STT_HINTS = {
```
.

swaroopvarma1 commented on app/ai/voice/agents/breeze_buddy/utils/language_utils/language_detector.py:
```diff
@@ -0,0 +1,202 @@
+GEMINI_LANGUAGE_DETECTION_MODEL = "gemini-2.5-flash"
+
+# Supported language codes (without Odia as Soniox doesn't support it)
+SUPPORTED_LANGUAGES = [
```
.

swaroopvarma1 commented on app/ai/voice/agents/breeze_buddy/utils/language.py:
```diff
@@ -0,0 +1,202 @@
+        return None
+
+
+def get_stt_hints_for_language(language_code: str) -> list[str]:
```
.

swaroopvarma1 commented on app/ai/voice/agents/breeze_buddy/agent.py:
```diff
@@ -259,14 +269,20 @@ async def run(self):
+        # 1. Explicit stt_language from template config (highest)
+        # 2. LLM-detected language from payload
+        # 3. Default (None → Soniox uses en,hi from static config)
+        if template and template.configurations and template.configurations.stt_language:
```
.

swaroopvarma1 commented on app/ai/voice/tts/sarvam.py:
```diff
@@ -44,21 +105,61 @@ def get_sarvam_language(language_code: Optional[str]) -> Language:
+    and switches the TTS language code before generating speech.
+    """
+
+    async def _switch_language_if_needed(self, text: str) -> bool:
```
.

swaroopvarma1 commented on app/api/routers/breeze_buddy/leads/handlers.py:
```diff
@@ -162,6 +154,17 @@ async def push_lead_handler(req: PushLeadRequest, current_user: UserInfo) -> Dic
+
+        # Add detected language code and human-readable name to payload for use during call
+        lead_payload["detected_language_code"] = detected_language_code
+        lead_payload["language_name"] = language_name
```
\

swaroopvarma1 replied:
detected_language_code -> language_code

swaroopvarma1 commented on app/api/routers/breeze_buddy/deprecated/leads.py:
```diff
@@ -127,6 +145,19 @@ async def trigger_order_confirmation(
             )
 
+        # Fetch template to check payload_based_language_selection setting
+        template_obj = await get_template_by_merchant(
```
.

narsimhaReddyJuspay commented:
@coderabbitai review the code

narsimhaReddyJuspay committed:
(feat): improve-multilingual-support
- added dynamic tts language switching
- address based language selection for stt and tt
- improved prompt for better multilingual support

swaroopvarma1 commented on app/ai/voice/agents/breeze_buddy/agent.py:
```diff
@@ -267,6 +279,20 @@ async def run(self):
+
+        # Check if language-aware prompting is enabled
+        # Enable if stt_language is set OR payload_based_language_selection is enabled
+        payload_based_selection = (
```
.

swaroopvarma1 commented on app/ai/voice/agents/breeze_buddy/agent.py:
```diff
@@ -414,6 +444,45 @@ async def run_pipeline():
             )
             await run_pipeline()
 
+    def _inject_language_rules(self, role_messages: list) -> list:
```
move this to a utils

swaroopvarma1 commented on app/ai/voice/tts/sarvam.py:
```diff
@@ -1,16 +1,76 @@
+            if 0x0041 <= code_point <= 0x005A or 0x0061 <= code_point <= 0x007A:
+                script_counts["english"] += 1
+
+    max_script = "english"
```
.

swaroopvarma1 commented:
@coderabbitai review

swaroopvarma1 commented:
@codex review
